### PR TITLE
Providing more control over hover label in plots

### PR DESCRIPTION
**Pull request by AlanRace**
I found that the label displayed at the cursor when hovering on a plot is only shown if either `show_x` or `show_y` were set to `true`. I wanted to be able to display the label regardless of whether the line guides were shown or not, and so added in the `show_hover_label` function to control this display.

I also found that `show_x` and `show_y` weren't very descriptive names for what functionality they were controlling (originally I thought it was related to the x-axis), so I have suggested a change in name to `show_hover_line_x`, but feel free to suggest something more fitting. Similarly with `custom_label_func`.

Finally, I modified the definition of `CustomLabelFunc` to enable a custom function to essentially recreate the default (this wasn't previously possible as it didn't have access to `show_x` and `show_y`. 

An example of a custom label
Custom label

### CODE REVIEW

File: egui/src/widgets/plot/mod.rs
Comment (Bromeon) [inline egui/src/widgets/plot/mod.rs:233]: This seems to be used in a few places, also in `bar.rs` and `box_elem.rs` -- maybe extract it to a function?
A few similar helper functions are already at the end of `items/mod.rs`.

File: egui/src/widgets/plot/mod.rs
Comment (Bromeon) [inline egui/src/widgets/plot/mod.rs:24]: Since `CustomLabelFuncRef` seems to be the prevalent type used everywhere, maybe we should think about a shorter/more precise name? `BarChart` and `BoxPlot` have an `element_formatter()` method, which customizes the displayed text when hovering over them, which is essentially the same concept.

Maybe something like `HoverFormatter`?

On the other hand, the type `dyn Fn(&HoverConfig, &str, &Value) -> String` seems to be used only once, so it can either be inlined or use a more verbose name `HoverFormatterFn` or so.

We are going to need a lot of these customization functions, so agreeing on a term like "formatter" could make sense.
Use cases:
* Hovering for bar charts and box plots (currently `element_formatter`, could be renamed to `hover_formatter`)
* Hovering for X/Y values (this PR, maybe `hover_formatter` makes clear it's the same idea)
* Labeling X/Y axis (I'm going to work on this, I'm thinking about `x_axis_formatter` + `y_axis_formatter`)

**Review by Bromeon — COMMENTED** [review]
Very nice! I also like the internal simplification (better naming + integrating to `PlotConfig`).
I have a few additional ideas to make things more intuitive/consistent 🙂

File: egui/src/widgets/plot/mod.rs
Comment (Bromeon) [inline egui/src/widgets/plot/mod.rs:224]: Why is this public? Someone who wants the default formatter would simply not call `hover_label_func()`.

**Comment by emilk** [comment]
It would be great if a demo of the new feature was added to `egui_demo_lib/src/apps/demo/plot_demo.rs`, and perhaps a screenshot added to the PR description

File: egui/src/widgets/plot/mod.rs
Comment (AlanRace) [reply egui/src/widgets/plot/mod.rs]: That's a good point. Having it public would have made the `plot_demo` code slightly easier to write, but that's probably too much of an edge case.

**Comment by AlanRace** [comment]
I have now updated with the comments above and added a new demo. A screenshot of the demo is also included in the original PR description.

File: egui/src/widgets/plot/mod.rs
Comment (emilk) [inline egui/src/widgets/plot/mod.rs:119]: Seems like it would make more sense to make this into a

```
pub struct HoverLine {
    /// Highlight values on the X axis by painting a vertical line.
    pub x: bool,
    /// Highlight values on the Y axis by painting a horizontal line.
    pub y: bool
}
```

**Review by emilk — CHANGES_REQUESTED** [review]
Looks great, but I think `HoverLine` makes more sense as a `struct` than an `enum`

**Comment by Titaniumtown** [comment]
This PR looks really cool! Would love to see this feature merged.

**Comment by emilk** [comment]
Coverted to a draft until work resumes!

**Comment by AlanRace** [comment]
Is this still necessary after merging https://github.com/emilk/egui/pull/1235?

**Comment by alexespencer** [comment]
I think it is still needed even though #1235 has merged as this can disable the hover tooltip completely?

**Comment by emilk** [comment]
`egui_plot` has recently been moved to its own repository, at https://github.com/emilk/egui_plot

This will hopefully speed up its development by having more reviewers and maintainers.

Please re-open this PR at https://github.com/emilk/egui_plot/pulls

See also:
* https://github.com/emilk/egui/issues/4705
* https://github.com/emilk/egui/pull/4828

**Pull request closed without merging**
